src/CoreWCF.Primitives/src/CoreWCF/Channels/SymmetricSecurityBindingElement.cs

```diff
@@ -0,0 +1,214 @@
+            if (ProtectionTokenParameters != null)
+            {
+                supportsClientAuthentication = supportsClientAuthentication || ProtectionTokenParameters.SupportsClientAuthentication;
+                supportsClientWindowsIdentity = supportsClientWindowsIdentity || ProtectionTokenParameters.SupportsClientWindowsIdentity;
+
+                if (ProtectionTokenParameters.HasAsymmetricKey)
+                {
+                    supportsServerAuthentication = ProtectionTokenParameters.SupportsClientAuthentication;
+                }
+                else
+                {
+                    supportsServerAuthentication = ProtectionTokenParameters.SupportsServerAuthentication;
+                }
+            }
+
+            return new SecurityCapabilities(supportsClientAuthentication, supportsServerAuthentication, supportsClientWindowsIdentity,
+                ProtectionLevel.EncryptAndSign, ProtectionLevel.EncryptAndSign);
+        }
+
+        internal override bool SessionMode
+        {
+            get
+            {
+                SecureConversationSecurityTokenParameters secureConversationTokenParameters = this.ProtectionTokenParameters as SecureConversationSecurityTokenParameters;
+                if (secureConversationTokenParameters != null)
+                    return secureConversationTokenParameters.RequireCancellation;
+                else
+                    return false;
+            }
+        }
+
+        internal override bool SupportsDuplex
+        {
+            get { return this.SessionMode; }
+        }
+
+        internal override bool SupportsRequestReply
+        {
+            get { return true; }
+        }
+
+        public override void SetKeyDerivation(bool requireDerivedKeys)
+        {
+            base.SetKeyDerivation(requireDerivedKeys);
+            if (_protectionTokenParameters != null)
+                _protectionTokenParameters.RequireDerivedKeys = requireDerivedKeys;
+        }
+
+        internal override bool IsSetKeyDerivation(bool requireDerivedKeys)
+        {
+            if (!base.IsSetKeyDerivation(requireDerivedKeys))
+                return false;
+
+            if (_protectionTokenParameters != null && _protectionTokenParameters.RequireDerivedKeys != requireDerivedKeys)
+                return false;
+
+            return true;
+        }
+
+       
+
+        public override T GetProperty<T>(BindingContext context)
+        {
+            if (context == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("context");
+
+            if (typeof(T) == typeof(ChannelProtectionRequirements))
+            {
+                AddressingVersion addressing = MessageVersion.Default.Addressing;
+#pragma warning disable 56506
+                MessageEncodingBindingElement encoding = context.Binding.Elements.Find<MessageEncodingBindingElement>();
```

Comment:
You can probably remove this warning as I think it applied to an internal build system code analyzer.